File: calculateTax.py
#Time: O(n)
#Space: O(1)
#Program ran on leetcode successfully

class Solution:
    def calculateTax(self, brackets: List[List[int]], income: int) -> float:

        tax = 0
        prev_limit = 0

        for b in brackets:
            limit = b[0]
            rate = b[1]
            tax += (min(limit, income) - prev_limit) * (rate / 100)
            prev_limit = limit

            if limit >= income:
                return tax
        
        return 0

File: calculate-tax.cpp
//Time - O(1)
//Space - O(1)
class Solution {
public:
    double calculateTax(vector<vector<double>> levels, double salary) {
        double remaining = salary, percent = 0.0, tax = 0.0;
        int prevLimit = 0;
        int slab = 0;
        
        while(remaining>0){
            percent = levels[slab][1];
            if(slab == levels.size()-1){
                tax = tax + remaining*percent;
                return tax;
            }
            double slabAmount = levels[slab][0]-prevLimit;
            tax = tax + min(remaining, slabAmount)*percent;
            prevLimit = levels[slab][0];
            remaining = remaining - slabAmount;
            slab++;
        }
        return tax;   
    }
};
